### ISSUE TITLE
API Refactorization

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#1392]
Layers: @jacquesqiao  should be done in 2017-2-23

- [调研]implement more layers in v2 API
  - issue #1402 
  - [100%] add cost function #1416 
  - [100%] add ParamAttr #1431
  - should be done in 2017-2-23
- [100%]specify data types of data layers
  - Merged https://github.com/PaddlePaddle/Paddle/pull/1386
  - Done
- [90%] add topology. to be discussed
    - https://github.com/PaddlePaddle/Paddle/pull/1400
    - should be done in 2017-2-23
- [80%]调通 RNN   @luotao1 + @jacquesqiao
   - should be done in 2017-2-23
     - 2017-2-23  单层已经调通，正在优化
     - 等待projection+mixlayer来实现多层嵌套的rnn。
- [80%] 实现projection+mixlayer   @qingqing01 + @jacquesqiao
   - 2017-2-23  方案讨论确定。

Data: should be done in 2017-2-27

- [100%]data feeder (from data reader to Argument) @qingqing01
  - https://github.com/PaddlePaddle/Paddle/pull/1374 
- [99%]ReaderCreator/ReaderDecorator @helinwang 
  - https://github.com/PaddlePaddle/Paddle/pull/1364
  - https://github.com/PaddlePaddle/Paddle/pull/1411
  - https://github.com/PaddlePaddle/Paddle/pull/1412
  - https://github.com/PaddlePaddle/Paddle/pull/1424
  - https://github.com/PaddlePaddle/Paddle/pull/1425

* Use sklearn/NLTK to get dataset, it should be complete before 2-25.
<s> - [20%]Data Package @wen-bo-yang @wangkuiyi
  * Data Package job in #1419
  * SRL data should be done in 2017-2-23
- [70%] paddle.data.caching @reyoung  
   * Data Fetcher
      * design #1394
      * implementation #1405
   * It seems that fetcher is not needed. Maybe we should fetch data independently.
   * Should be done in 2017-2-23 </s>

Activations  should be done in 2017-2-23
- [100%]C++ getLayerOutput() and Expose To SWIG and update python getLayerOutput @gangliao https://github.com/PaddlePaddle/Paddle/pull/1353
<s> - [100%] Paddle Docker CPU and GPU
https://github.com/PaddlePaddle/Paddle/pull/1443

Event: @reyoung  should be done in 2017-2-23

- [70%]Get Strong Typed Metric From `paddle::Evaluator`: 
  * also, there are several tiny fixes merged to develop. #1371 #1372 #1373 
  * issue #1389  
  * PR #1375
  * Should be done in 2017-2-23

- [没开始]补全Event？
- [调研]IPython 画图？

  - https://github.com/PaddlePaddle/Paddle/issues/1415

- [没开始]其他的Demo？

**Comment by qingqing01** [comment]
从[semantic_role_labeling](https://github.com/PaddlePaddle/Paddle/tree/develop/demo/semantic_role_labeling)出发，写API训练中，一些想到的问题：

## 网络配置
 -  ParameterAttribute、xx_projection等其他如何引入包
  这些目前在： python/paddle/trainer_config_helpers，还继续 import 这些用嘛？

## 训练
-  parameters
   初始化问题：C++还是Python？如果是python需要提供常用初始化接口吧？

- optimizer
  其他demo需要完善python/paddle/v2/optimizer.py内容，目前只有Adam，Adamax。
  
- L2Regularization
  如何设置？ 6 / 8 demo有正则

- Testing during training
   event_handler需要支持
  
## 其他

- 疑问：
网络连接如下：

``` 
          B_cost_layer：是cost层, paddle.parameters.create(B_layer)即可训练
        / 
A_layer
        \ 
         C_layer  和 B_layer共享参数，testing用。
 
```
按照训练topology=B_cost_layer即可，topology=[B_cost_layer, C_layer]吗? 否则的话testing如何拿到C_layer的输出？
```
    trainer.train(train_data_reader=train_reader,
                  topology=[B_cost_layer, C_layer],
                  ...)
```

**Comment by wangkuiyi** [comment]
@qingqing01 非常感谢指出这些问题。

> ParameterAttribute、xx_projection等其他不在python/paddle/v2中的如何引入包。这些目前在： python/paddle/trainer_config_helpers

这个问题我不了解源码，是不是应该和其他更了解的同学讨论一下，给出一个解法？

> parameters初始化问题：C++还是Python？如果是python需要提供常用初始化接口吧？

这个确实需要接口。我记得之前 @reyoung 给出的一个设计里考虑了这个问题。我还记得后来 @reyoung 把结构设计修改成 `Parameters.set/get` 了。

> optimizer：其他demo需要完善python/paddle/v2/optimizer.py内容，目前只有Adam，Adamax。

这是因为我们的book里用到了除了Adam和Adamax之外的其他optimization algorithms了吗？如果是，那么貌似确实得尽快完备；否则，是不是可以稍后完备？

> L2Regularization如何设置？ 6 / 8 demo有正则

这个是不是可以在train函数里增加一些参数来设置？

> Testing during training：event_handler需要支持

在我脑海里，我理解你给出的例子可以这么描述：

```python
a_layer = paddle.layer.SomeLayer()
b_cost = paddle.layer.SomeCost(a_layer)
c_cost = paddle.layer.SomeOtherCost(a_layer)

parameters = paddle.parameters.create([b_cost, c_cost])

def event_handler(event):
    if isinstande(event, paddle.event.EndIteration):
        print test(c_cost,
                   parameters,
                   paddle.dataset.mnist.testing_data())
        
paddle.train(b_cost,
             parameters,
             paddle.dataset.mnist.training_data(),
             event_handler)
```

请看这样可以吗？

**Comment by reyoung** [comment]
This issue has been done.
